src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs

```diff
@@ -1355,6 +1358,24 @@ private void ConfigureWarningsAsErrorsAndMessages()
             }
         }
 
+        private void ConfigureKnownImmutableFolders()
+        {
+            ProjectInstance project = _requestEntry?.RequestConfiguration?.Project;
+            if (project != null)
+            {
+                //example: C:\Users\USER_NAME\.nuget\;C:\Program Files\dotnet\sdk\NuGetFallbackFolder
+                FileClassifier.Shared.RegisterImmutableDirectories(project.GetPropertyValue("NuGetPackageFolders")?.Trim());
+                // example: C:\Windows\Microsoft.NET\Framework\v4.0.30319\
+                FileClassifier.Shared.RegisterImmutableDirectories(project.GetPropertyValue("MSBuildFrameworkToolsPath32")?.Trim());
+                // example:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\
+                FileClassifier.Shared.RegisterImmutableDirectories(project.GetPropertyValue("MSBuildFrameworkToolsPath64")?.Trim());
+                // example: C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.7.2
```

Comment:
These are defined in the toolset. Should they be registered there instead of here?

Reply:
Toolset seem to be better place. I think about putting it bellow this line:
https://github.com/dotnet/msbuild/blob/9bcc06cbe19ae2482ab18eab90a82fd079b26897/src/Build/Definition/ToolsetReader.cs#L423

Reply:
done